### ISSUE TITLE
Is there a way to keep using clink v0.4.9 on Windows 11 without upgrading Clink?

### ISSUE THREAD

**Issue by ben-may** [chrisant996/clink#174]
When using clink on windows 11 we find that it does not seem to properly hook things, and I also don't see the lua scripts being run either. Not sure what is going on TBH. Are you able to run clink properly on Windows 11? I am using v0.4.9.
I know that its an old version of clink but I'm not in a position to be able to wholesale upgrade clink.

**Comment by chrisant996** [comment]
> When using clink on windows 11 we find that it does not seem to properly hook things, and I also don't see the lua scripts being run either. Not sure what is going on TBH. 

Clink v0.4.9 made some unsafe assumptions about internals of Windows.  The assumptions started failing in Win10 about a year ago.  v0.4.9 is very old and is not supported.

> Are you able to run clink properly on Windows 11?
 
Versions of Clink from the past year or so work (v1.x), i.e. from this chrisant996/clink repo.

> I am using v0.4.9. I know that its an old version of clink but I'm not in a position to be able to wholesale upgrade clink.

If you're concerned about compatibility, I've gone to great lengths to make current versions of Clink compatible with v0.4.9.  A few key binding strings might be different, but other than that it's highly likely that everything will just work, and you'll get lots of new features and bug fixes.  Cmder has replaced their Clink v0.4.9 with recent Clink versions, and they have proven to be compatible so far.

**Comment by ben-may** [comment]
I've taken the latest release of clink now and have been able to successfully upgrade to it!  thanks for the backwards compatibility that has been done.  And yes it seems to work fine in Windows 11.

**Comment by ben-may** [comment]
Actually @chrisant996 I am having some issues with lua function calls that used to be in clink.lua, which is now empty.  I know that the lua files are now compiled into clink itself now, but did the lua functions get renamed etc.?  I'm having some of our existing lua extensions calling functions like clink.split() and clink.match_files() etc. and those now fail since the function isn't found anymore.  Is there any docs on where these functions went or how to search-n-replace in my lua files to the new function calls?

**Comment by chrisant996** [comment]
Drat.  I caught all the Lua API functions implemented in C++, but I missed a few that were implemented in Lua in clink.lua itself.

For context:  Martin did a massive rewrite for his v1.0 that was never published, and he removed clink.lua completely.  The rewrite had no backward compatibility yet when he abandoned the project (I do hope he's ok).  I had to manually track everything down and add extensive backward compatibility shims, and I missed a few functions written in Lua.

**I'll publish an update with those this evening.**

In the meantime, you could maybe copy the needed functions from the old clink.lua file (don't copy the whole file, as that would break things -- only copy/paste specific functions).  `clink.split()` is an easy one.  However `clink.match_files()` might be more involved than a straight copy/paste.

**Comment by ben-may** [comment]
Yeah in order to unblock myself I copied clink.split() and clink.quote_split() into our own lua files, but match_files() I haven't tried yet.  I just figured that since you were saying there was good backwards compatibility with 0.4.9 I was thinking I was doing something wrong or missed an upgrade guide since I had hit the 3rd function that was missing 😄 
Anyhow any pointers or quick ports of match_files() would be great if you have them.  I'm happy to just paste that into our lua files for now, until there is an official version of clink I can pull to fix that.

**Comment by chrisant996** [comment]
Would it be possible to share the Lua file that calls clink.match_files?

The broader context may help find a quick workaround.

**Comment by ben-may** [comment]
match files also calls match_count() so its pulling a string... let me get you some code to look at

**Comment by ben-may** [comment]
Here is an example block of code that is calling into match_files:
`function mylua.data_match_generator(word)
	local special_folder = clink.get_env("special_folder")
	if not special_folder then
		return {}
	end
		
	special_folder = special_folder.."/folderA/"
	
	local cwd = clink.get_cwd()
	clink.chdir(special_folder)
	
	clink.match_files(word.."*", true, clink.find_dirs)
	clink.match_files(word.."*xml", true)
	
	clink.matches_are_files()
	return {}
end`

**Comment by chrisant996** [comment]
It looks like copying `clink.match_files()` will work, if you also add the following stub function:

```lua
function clink.match_count()
    return 0
end
```

Do you have any Lua code that uses clink.match_count() or clink.get_match()?  

This discussion has uncovered an interesting compatibility detail.  In the past, Clink's match generators gave read access to the matches as they are being generated.  That was removed in Martin's v1.0 many years ago, which wasn't released at the time.  Now generators only have write access, and there is no way to query the number of matches or access individual matches.  I think that's the right abstraction to have there.

I want to avoid granting back-compat access to clink.match_count() or clink.get_match().  Those aren't safe, and they would be limiting to Clink and to scripts, and they may present a ripple effect of incompatibilities if they were to be added.

I'm pretty sure that anywhere that uses these can be fairly easily adjusted to not need them.  I'm happy to help with that.  You can send me a .zip with all the scripts you want reviewed/updated if that would be helpful.

**Comment by ben-may** [comment]
I can try adding that stub, although, won't this math in clink.match_files() be bad then?
```
   -- Match them.
    local count = clink.match_count()

    for _, i in ipairs(glob) do
        local full = base..i
        clink.add_match(full)
    end

    return clink.match_count() - count
```

**Comment by ben-may** [comment]
adding the stub seems to have stopped it from crashing, but the completion I'm getting is not appending the trailing '\' in the path completion on some of my commands now.  Any ideas?

**Comment by ben-may** [comment]
I'm also not sure I can give you a zip of all of our lua, since I don't think I'm allowed to share that publically.

**Comment by ben-may** [comment]
oh weird the trailing '\\' is showing up now, but the returned results are still wrong, missing a ton.

**Comment by ben-may** [comment]
Another difference I noticed is that before the tab-completion options that get listed were coloured now they are not, I'm not sure where that was configured/setup before.

**Comment by chrisant996** [comment]
> I can try adding that stub, although, won't this math in clink.match_files() be bad then?

The math will be wrong. But nothing uses the return value, so no harm.

> adding the stub seems to have stopped it from crashing, but the completion I'm getting is not appending the trailing '' in the path completion on some of my commands now. Any ideas?

Can you clarify? Directories normally don't get a " appended (not in bash, clink v0.4.9, or clink v1.x). Only files do.

**Comment by ben-may** [comment]
sorry a backslash, '\\' on folders.  
and yeah my tab completion options were always listed in purple, now they are just uncolored.

**Comment by ben-may** [comment]
I found another missing function disable_file_matching()

**Comment by chrisant996** [comment]
I'll take a closer look tonight, and will hopefully be able to post an update tonight with all reported issues fixed.

Keep 'em coming!

And if there's any way I can access the Lua scripts, then I could save you from the unexpected troubleshooting costs.  I know you said you can't do a wholesale update right now, but that's the only option for continuing to use Clink.  I'm willing to take on that cost myself, since it also benefits the broader community.

(I can sign an NDA and delete the scripts immediately afterwards, if that would help.)

> sorry a backslash, `\` on folders.

Backslash on folders is normally controlled by the `mark-directories` option in the .inputrc file.

Quick possibilities:
- Check the value of `mark-directories` in the .inputrc file.  The default is 1 (`on`), but if clink v0.4.9 didn't respect that option and the .inputrc file turned it off, then that could explain the difference in behavior upon updating.
- Match generators have the ability to suppress adding the backslash on directories.  Maybe there is an additional back-compat shim missing somewhere, but I would anticipate the opposite problem if that were the case (i.e. always adding `\` even when the generator says not to).
- Some other bug (e.g. recently there was a bug in the append-to-match logic for symlinks, but that should be fixed.)

> my tab completion options were always listed in purple, now they are just uncolored.

Not sure I follow what that refers to.  Completion colors are whiz-bang now:  https://chrisant996.github.io/clink/clink.html#completion-colors

Do you have match display filters that apply purple to some of the matches?  Match display filters that apply coloring should be working (they work for other things such as the [clink-completions](https://github.com/vladimir-kotikov/clink-completions) repo).

I can't think what else "purple completion options" could refer to.
Would a screenshot of the old purple completion options be possible?

(Can email privately for privacy, to the address in my github profile.)

**Comment by ben-may** [comment]
Thanks for all the extra details, I'll keep posting issues as I find them.  And I've already decided that the only way forward is to take this latest version of clink and make it work, v0.4.9 is going to be too much hassle to sort out, so I will proceed with v1.2.35.fa2a3e and hopefully with any fixes you post we'll be fixed up and ready to go! Thanks for all the support and help!  without it I may have just decided to abandon clink :)

**Comment by ben-may** [comment]
![image](https://user-images.githubusercontent.com/16196367/135535490-043565cd-f093-4ab8-a463-3ca910648a21.png)
Here is a screenshot of the purple tab-completion colours

**Comment by ben-may** [comment]
I found it, we used to use this setting:
`clink.match_colour`
That setting doesn't exist anymore I found this in the code:
```
    // `match_colour` is no longer a setting; use `colored-stats` in the inputrc
    // file and `set LS_COLORS` to set the colors.  Also certain `color.*` Clink
    // settings.
```

**Comment by ben-may** [comment]
I have been getting a lot of these also:
`warning: replacing arglink for 'uninstall' in ...\,mylua.lua:418 (see log file for details).`
looks like all of the arg parsers I have from v0.4.9 are using deprecated ways of setting these up?

**Comment by chrisant996** [comment]
Yes but the deprecated way still works.

It looks like multiple argmatchers are being registered for the same command.

The fix is probably to track down which one should be kept, and remove the others.  I shouldn't make Clink be silent about this, because even v0.4.9 doesn't work entirely properly in these cases, and it's (very) ambiguous what "should" happen (Clink can't make a reasonable guess).

The only case where v0.4.9 is expected to work properly is if the subsequent re-registrations exactly match what was already there (in which case it's inefficient, and also sets you up for problems later if any of the copies of the argmatcher definitions are changed).

**Comment by chrisant996** [comment]
> I found it, we used to use this setting:
> `clink.match_colour`
> That setting doesn't exist anymore I found this in the code:
> ```
>     // `match_colour` is no longer a setting; use `colored-stats` in the inputrc
>     // file and `set LS_COLORS` to set the colors.  Also certain `color.*` Clink
>     // settings.
> ```
> I've experimented with readline's colored-completion-prefix and colored-stats options which give something close but not quite the same.

Oh right, sorry, that should have clicked for me.

It's true, Readline's new whiz-bang completion coloring assumes you want to see lots of color coding.

I think it's possible to make all completions be the same color, but I'm not positive. I'll look this evening.

**Comment by ben-may** [comment]
> Yes but the deprecated way still works.
> 
> It looks like multiple argmatchers are being registered for the same command.
> 
> The fix is probably to track down which one should be kept, and remove the others. I shouldn't make Clink be silent about this, because even v0.4.9 doesn't work entirely properly in these cases, and it's (very) ambiguous what "should" happen (Clink can't make a reasonable guess).
> 
> The only case where v0.4.9 is expected to work properly is if the subsequent re-registrations exactly match what was already there (in which case it's inefficient, and also sets you up for problems later if any of the copies of the argmatcher definitions are changed).

I've worked through these and seem to have removed all the warnings, the way things were setup was it was continuously re-registering the same parser by the same name, so it needed to always merge, I instead just used the newer api to get the existing parser and add to it instead.

**Comment by chrisant996** [comment]
The following should make all completions use the same color.
I'll use magenta as the example, since that's what you had configured for v0.4.9.

Add this to the .inputrc file:
```inputrc
set colored-stats on
```

Set LS_COLORS in the environment or in a batch script:
```cmd
set LS_COLORS=fi=35:di=35:ex=35:no=35:mi=35:or=35
```

Set some Clink color settings:
```cmd
clink set color.cmd ""
clink set color.doskey ""
clink set color.filtered magenta
```

**Comment by chrisant996** [comment]
> I've worked through these and seem to have removed all the warnings, the way things were setup was it was continuously re-registering the same parser by the same name, so it needed to always merge, I instead just used the newer api to get the existing parser and add to it instead.

I recommend to assign the parser into a variable, and use the variable when wanting to modify or reference the parser.

> backslash `\` on folders

Did you get this one working?
My first guess would be that the .input rc has `set mark-directories off`, and it would need to be `on`.

> _... the various missing functions ..._

I've added several compatibility shims.  Some things don't have exact analogs anymore.

I ran out of time tonight, and might not be able to prioritize these further until early next week.

But here is a pre-release build that should work a bit better.
[clink.1.2.36.4e0d4b.zip](https://github.com/chrisant996/clink/files/7265355/clink.1.2.36.4e0d4b.zip)

For things that are not supported anymore, it will report warnings.
The warnings are suppressed when %CLINK_COMPAT_WARNINGS% is 0.
The warnings include stack traces when %CLINK_COMPAT_WARNINGS% is 2.

**Comment by ben-may** [comment]
> The following should make all completions use the same color. I'll use magenta as the example, since that's what you had configured for v0.4.9.
> 
> Add this to the .inputrc file:
> 
> ```inputrc
> set colored-stats on
> ```
> 
> Set LS_COLORS in the environment or in a batch script:
> 
> ```batchfile
> set LS_COLORS=fi=35:di=35:ex=35:no=35:mi=35:or=35
> ```
> 
> Set some Clink color settings:
> 
> ```batchfile
> clink set color.cmd ""
> clink set color.doskey ""
> clink set color.filtered magenta
> ```

This worked thanks!

**Comment by ben-may** [comment]
I may have misunderstood what :nofiles() function on the parsers does, I assumed it meant that no files were returned in the completion results, but I'm seeing many circumstances where they are still returned along with my other completion.  Perhaps nofiles() is not the same as disable_file_matching() in some way?

**Comment by chrisant996** [comment]
> I may have misunderstood what :nofiles() function on the parsers does, I assumed it meant that no files were returned in the completion results, but I'm seeing many circumstances where they are still returned along with my other completion results.  Perhaps nofiles() is not the same as disable_file_matching() in some way? 

Maybe you could post a sample generator with sensitive strings redacted, so that I can see what is happening? Otherwise it's very difficult to make guesses.

There are several ways that `disable_file_matching()` could be used, and maybe one of them has additional side effects that I didn't recognize.

**Comment by ben-may** [comment]
I am seeing quite a few places where I have to re-write my argument parsers, and things that seem like they are supposed to work but don't.  Let me see if I can give you a few examples, so you have something to work with.  I think that part of my problems I'm seeing is that the :loop() behaviour has changed. Instead of looping on the last arg that was added to the parser, it assumes to loop back to arg 1, so I am having to go fix a bunch of them now to pass an actual argument index into the loop() function.  I fear this is going to be a lot of work to rework all of my argument parsers 😢

**Comment by ben-may** [comment]
I also noticed that flags cannot come after args, wheras in old clink they could and there is a couple places where that is important for us.  A use case is things like passthrough arguments.

**Comment by chrisant996** [comment]
There is something interesting or unusual about your argmatchers, so I will need to see them to figure out how to make them work without modification.

I don't want you to have to rewrite your stuff.

It's entirely possible that Martin totally changed how loop works, I don't know offhand. Will dig deeper early next week, esp. if a sample argmatcher is available.

Thanks!

**Comment by chrisant996** [comment]
> I also noticed that flags cannot come after args, wheras in old clink they could and there is a couple places where that is important for us.  A use case is things like passthrough arguments.

Something is wrong. Flags are not positional and can be intermixed with args. But a linked argmatcher naturally comes with its own separate collection of args.

**Comment by ben-may** [comment]
I will try and get you some sanitized arg parsers to test out and see if you can see similar behaviour on your side so you can more easily see what I'm seeing.

**Comment by ben-may** [comment]
Here is an example:

```
local function args_func(word)
	return {
		"-D:Aoption",
		"-D:Boption",
		"-D:Coption"
	}	
end

local passthroughparser = clink.argmatcher()
passthroughparser:loop()
passthroughparser:set_arguments( { args_func } )

local testparser = clink.argmatcher()
testparser:nofiles()
testparser:set_flags("--" .. passthroughparser)
testparser:set_arguments({"arg1", "arg2"})
```
This parser doesn't let me access the '--' passthrough flags if I first have added an "arg1" or "arg2" argument.  It also exhibits an odd bug when the ':' character is in the list of arguments the tab-completion gets messed up and doesn't properly show the options and also repeats -D: if you press tab again.

**Comment by ben-may** [comment]
one thing that keeps tripping me up is that :set_arguments() before used to just append, now it clears and sets it to what you pass in as a parameter.  I understand that set() methods usually throw away what was there before, but I'm finding that in previous clink set_arguments() was used to append.  So perhaps we can change the compatibility shim to just append instead of clear it first?

**Comment by chrisant996** [comment]
I'll look at set_arguments etc. I thought there were both set_ and add_ but maybe I misread.

The `-D:Aoption` issue is probably a bug in the new `:` and `=` feature for flags:

Now you can use a linked parser with the flag, e.g.

```lua
"-D:" .. clink.argmatcher():addarg("Aoption", "Boption")
```

And then if you have tons of possibilities they don't all show up when using `list-completions` (e.g. Alt+=).

I probably got something wrong with handing that when there's no linked argmatcher.  (Though I have some written each way for testing purposes, specifically to help catch such mistakes. 🤷‍♂️)

**Comment by ben-may** [comment]
Ah that would work for porting! thanks!

**Comment by ben-may** [comment]
yeah that seems to work around the problem for that specific case, but I still cannot get the flags to be able to be used after args, it only lets me use them before args :(

**Comment by chrisant996** [comment]
If I find time this weekend I'll look. Otherwise early next week.

**Comment by chrisant996** [comment]
> one thing that keeps tripping me up is that :set_arguments() before used to just append, now it clears and sets it to what you pass in as a parameter. I understand that set() methods usually throw away what was there before, but I'm finding that in previous clink set_arguments() was used to append. So perhaps we can change the compatibility shim to just append instead of clear it first? I suppose the same would be true for set_flags()

In Clink v0.4.9 `:set_arguments()` first explicitly sets the args to empty, and then adds.

Here is the source code for set_arguments in v0.4.9.

```lua
local function parser_set_arguments(parser, ...)
    parser.arguments = {}
    return parser:add_arguments(...)
end
```

And when creating a parser, it did the following:

```lua
    parser.set_arguments = parser_set_arguments
    parser.add_arguments = parser_add_arguments
```

I completely believe you that the end result isn't working the same.  But with respect to set_arguments, it used to clear the arguments in v0.4.9, and it clears the arguments in current Clink versions.

So something else is going on.

**Comment by chrisant996** [comment]
> yeah that seems to work around the problem for that specific case, but I still cannot get the flags to be able to be used after args, it only lets me use them before args :(

Can you share a specific example command line, for the example parser you shared?

Also, notice that the example uses `set_arguments( { args_func } )`.
It uses set **arguments** to set **flags**.

That may be the root cause for why things are going wrong.
In Clink v0.4.9, mismatches like that didn't fully work correctly, but they also sort of worked _**if**_ the parser did not get told what the flag character was (e.g. `-`).

Newer Clink automatically figures out the flag character(s), and fixed a bunch of bugs.  However, that pre-existing mismatches in parsers that previously had been sort-of-working may stop working (i.e. they may start behaving how they've been coded).

I can try to take a look at how to reproduce previous bugs and malfunctions.  I think it's reasonable to expect certain kinds of bugs to be replicated moving forward, when using the deprecated APIs, but not when using the new APIs.  Fortunately it's pretty easy to distinguish at runtime which set of APIs is being used.

However, some quirks/bugs/malfunctions may be nearly impossible to replicate due to the major internal rewrite that Martin did on Clink several years ago, which didn't get release until a year ago after I did my best to finish it, and also extend it.

**Comment by chrisant996** [comment]
# Update Sun 10/3

I ran out of time for today, and there's a regression in the test suite, so I won't post another pre-release build until after I've fixed the test suite regression.

### Changes
Here are the fixes that are queued up.

- The `-D:Aoption` / etc stuff should work without modification now; i.e. what you were using with v0.4.9.  (Or you can use the new linked parser support, instead.)
- The `clink.suppress_char_append()` and `clink.suppress_quoting()` should work without modification now.
- Several other compatibility shims added.
- For things that are not supported anymore at all, it will report warnings.
    - The warnings are suppressed when %CLINK_COMPAT_WARNINGS% is 0.
    - The warnings include stack traces when %CLINK_COMPAT_WARNINGS% is 2.

### Can't Reproduce
I'm not able to reproduce the following issues.

Please share a sample parser + specific command line for each of the follow, so that we can track down that's going on:

- You've stated that in v0.4.9 `clink.disable_file_matches()` prevents file matches, but that `:nofiles()` in current versions does not.  I can't reproduce that; `:nofiles()` is preventing file matches in my testing, in the many places that Clink internally uses it, and in my own parsers.
- You've stated that in v0.4.9 `set_arguments` added arguments without clearing existing arguments.  That's inaccurate.  It always cleared existing arguments.
- You've stated that in v0.4.9 `:loop()` only loops on the most recent argument by default.  That's inaccurate.  It always started over at arg 1 in the associated parser.

**Comment by chrisant996** [comment]
Here is the pre-release build to go along with the previous post:
[clink.1.2.36.495d90.zip](https://github.com/chrisant996/clink/files/7276707/clink.1.2.36.495d90.zip)

**Comment by ben-may** [comment]
Thanks a ton @chrisant996 !  I'm not sure what is going on with set_arguments() tbh.  and loop() I think you are right, that was a mis-diagnosis of something else that was going wrong. I see instances in our parsers where loop() was being used with an integer to loop back to a specific place.

I will try out your pre-release, I am building from source though, so I'll just grab head of master.

**Comment by ben-may** [comment]
So thank-you for the compatibility shims, that helps with the need to search and replace things.
I have this sample/test parser and I had a few questions about its behaviour:
```
local test_parser = clink.argmatcher()
test_parser:addflags("-h", "--help", "-show_output", "-flag1", "-optionB")
test_parser:addarg({ "arg1", "arg2",  "something", "else" })
test_parser:nofiles()


local testB = clink.argmatcher()
testB:addflags("-flagA", "-flagB", "-optionC")
testB:addarg({ "hey", "this",  "isa", "test" })
testB:nofiles()

--------------------------------------------------------------------------------

local fb_parser = clink.argmatcher("t")
fb_parser:addarg(
	{
		"test_parser" .. test_parser,
		"test-B" .. testB 
	}
)
```
1. Using "nofiles()" here I would assume that if I did `t test_parser arg1 <tab>` that it should not return any file results, or do I not understand how nofiles() is supposed to work
2. I cannot seem to pass flags after any arguments still.  For example if I do `t test_parser arg1 -<tab>` then it does not show me any flag completion.
3. It seems that if tab-completion has a '_' and a '-' character in the same place trying to tab-complete on that character erases the character.  For example if I do `t test_<tab>` or `t test-<tab>` then it will erase the '_' or '-' character.

**Comment by chrisant996** [comment]
> 1. Using "nofiles()" here I would assume that if I did `t test_parser arg1 <tab>` that it should not return any file results, or do I not understand how nofiles() is supposed to work
> 2. I cannot seem to pass flags after any arguments still.  For example if I do `t test_parser arg1 -<tab>` then it does not show me any flag completion.
> 3. In old clink listing several {} inside addarg() (like in "test_parser" above) would give you multiple argument positions, now it seems to concatenate them together into a single argument position.  We have several of these so I'll need to go fix them.

This is a porting mistake.  Clink is doing exactly what it's been told to do.  And if you combine the two tables inside `add_arguments` and load that into v0.4.9, it exhibits the same problems.

Details:

These are all the same issue.  It looks like `add_arguments` was replaced with `addarg`.  But they are not semantically equivalent, so just renaming won't necessarily produce the same behavior.
- `:addarg(...)` is singular (no trailing `s`) -- it adds all of `...` as _**one**_ argument position.
- `:add_arguments(...)` is plural (trailing `s`) -- it adds one argument position _**per table**_ passed to it.

Either don't apply the rename, or make the dependent semantic changes as well as renaming.

As written, the `test_parser` has only one argument, with 4 values.  Once a linked parser runs out of arguments, it is done; it gets popped off the parser stack.

- Re: 1 -- `nofiles()` was used in the `test_parser` parser.  But that parser is not active anymore because it ran out of arguments.  And the `fb_parser` didn't use `disable_file_matching()` or `nofiles()`, so it still produces file matches.  That looks like an oversight in the original parser, maybe?
- Re: 2 -- The `test_parser` only has one argument, and it ran out of arguments.  So it's not the active parser at that point.  The `fb_parser` is what's active after the `arg1`.  And `fb_parser` has no flags.
- Re: 3 -- Yes.  `addarg` is documented as adding one argument.  `add_arguments` adds one argument per table passed to it.  They behave differently.  Either don't rename the calls, or rename them and also update them according to the new usage model.

> 4. It seems that if tab-completion has a '_' and a '-' character in the same place trying to tab-complete on that character erases the character.  For example if I do `t test_<tab>` or `t test-<tab>` then it will erase the '_' or '-' character.

This is a regression introduced in v1.1.1:
- Clink v0.4.9 always used Readline's completion system.
- In v1.0.0 Martin changed it to never use Readline's completion system.  That created very many problems.
- In v1.1.1 I changed it back to always use Readline's completion.
- Somehow along the way, the longest common prefix turned into the _literal_ longest common prefix, instead of treating `-` and `_` as equivalent.

I'll fix that tonight.

A couple of others have reported the same, but it wasn't clear to me until now that v0.4.9 retained the `-` or `_` in the longest common prefix.

**Comment by ben-may** [comment]
Thanks for the detailed response!  the flag behaviour is definitely different between clink 0.4.9 and the latest version, since If I use the same lua scripts under 0.4.9 the flags can be added after the arguments, and the parser still seems to be active, so maybe this is a behaviour change? I have corrected the arguments now though so that they just call addarg() multiple times instead of listing them.

**Comment by ben-may** [comment]
So on the add_arguments() -> addarg() are you considering that something that just needs to be adjusted during porting?  
Like you said previously there seemed to just be set_arguments() and add_arguments() and those were replaced with addarg() which does not treat multiple tables as multiple arg positions.

**Comment by ben-may** [comment]
I ended up re-writing my generator and use the built in clink.dirmatches() and clink.filematches() functions, this gives me the trailing slash as I want, as well as does not have the space appended.

**Comment by chrisant996** [comment]
> the flag behaviour is definitely different between clink 0.4.9 and the latest version, since If I use the same lua scripts under 0.4.9 the flags can be added after the arguments, and the parser still seems to be active, so maybe this is a behaviour change?

Oh, you're quite right, there is a change:
- Clink v0.4.9 has a bug, where it isn't able to pop out of nested parsers.
- Clink v1.0.0 fixed that bug (as part of Martin's big rewrite), and the side effect is the behavior does change in that edge case.

I can make Clink replicate that bug when the old `clink.arg.new_parser()` API is used to create a parser.

If you don't want to wait for that, you could achieve the old behavior by adding an empty arg and looping at the end:
```lua
local fb_parser = clink.arg.new_parser()
fb_parser:add_arguments(
	{
		"test_parser" .. test_parser,
		"test-B" .. testB 
	}
)
fb_parser:add_arguments({}) -- add an empty arg slot
fb_parser:loop(2) -- loop on the empty arg slot

clink.arg.register_parser("t", fb_parser)
```

> So on the add_arguments() -> addarg() are you considering that something that just needs to be adjusted during porting? Like you said previously there seemed to just be set_arguments() and add_arguments() and those were replaced with addarg() which does not treat multiple tables as multiple arg positions.

Right.  There is no longer a way to _remove_ arguments (apart from the deprecated set_arguments), and the new API as described in the docs is meant to be used in either of these ways:

```lua
clink.argmatcher("somecommand",
    {"arg1A", "arg1B"},
    {"arg2A", "arg2B"}
)

clink.argmatcher("somecommand")
    :addarg("arg1A", "arg2B")
    :addarg("arg2A", "arg2B")
```

For whatever it's worth, I had no hand in designing the old API or the new API.  I've read the commit history and the source code, and the new API seems to be designed with intent, and seems easier to use (and read!) than the old API, so I'm pretty happy to let the new API be what it is.

> In old clink `find_files()` and `find_dirs()` do seem to return a trailing path separator, so in new clink if I'm still using find_files and find_dirs() now I don't get a trailing slash.

I think you mean that listing or inserting completions generated via `find_files()` are not including trailing path separators in current versions of Clink.  I'm able to reproduce that.

The diagnosis of how/why is inaccurate, though.  For example, the two cited functions did not (and still do not) include trailing path separators.  There is a lot of code in between `find_files()` and listing or inserting a match.  The code in between is how spaces and slashes and percent signs and whatnot show up (or not) when listing or inserting matches.

Add this code to a .lua file and launch v0.4.9 if you would like to observe for yourself that `find_files()` did not include trailing separators.

```lua
for _,s in ipairs(clink.find_files(os.getenv("USERPROFILE").."/*")) do
    print(s)
end
```

> Also if I hack the code to add the slash (IE: disable the backcompat line in os_api.cpp) then the slashes are added properly, however then there is a space inserted after, so further tab-completion on the path starts over again since it appended a space. Is there a setting for that?

That adds slashes, but causes other problems.  The issue is that Clink is not realizing they are directories.  Thus it doesn't insert the trailing path separator (which was not part of the match text in v0.4.9) and it does insert a trailing space.  Adding a slash in the match text is just a slash in the match text -- it doesn't make Clink realize the match refers to a directory.

I'll fix the issue, and the commit diffs will hopefully help it make sense what's going on.

**Comment by chrisant996** [comment]
669bd83c1be0eb24b8d27bf492dc9ef050e4b8fd fixes the omitted trailing slashes when using the deprecated APIs for adding matches.

**Comment by chrisant996** [comment]
There are multiple subtle bugs impacting popping out of nested linked parsers.  I hadn't noticed that Martin hadn't gotten that working correctly yet.  It will take a few days to sort out how that should work.

In the meantime, as mentioned above (two posts back), you can force a parser to dead-end (as always happened in v0.4.9) by adding an empty arg slot and looping on that arg slot's index.

**Comment by chrisant996** [comment]
Right, that won't work.
Because it's not using the empty arg + loop() workaround suggestion.

What happens when you apply the suggested workaround?
It works just like v0.4.9 for me, so if it doesn't for you then I wonder if a mistake occurred while applying it?

> I think I have everything working as I'd expect except the flags, they still cannot be used after the args like in clink 0.4.9. Do you have any suggestions on how to acheive that?

I gave a suggestion how to achieve that.
And you said the suggestion works.
And in the next sentence (above), you said the suggestion doesn't work.
And then asked if I have any suggestions how to make it work.

So then I'm confused:  you must have tried to apply the workaround, because you said it works, but then you said it doesn't work.  The parser inside my head becomes stuck and can't figure out how to make sense of the statements.  😵

Here is the suggestion in action, applied to the sample parser:

```lua
    local test_parser = clink.argmatcher()
    test_parser:addflags("-h", "--help", "-show_output", "-flag1", "-optionB")
    test_parser:addarg({ "arg1", "arg2"})
-- IMPORTANT!  The following must be a second addarg() to add it as a second arg slot.
    test_parser:addarg({ "something", "else" })
-- IMPORTANT!  Add an empty arg.
    test_parser:addarg({})
-- IMPORTANT!  Loop on the empty arg.  The index must exactly match the actual index above.
-- If you used loop(3) the way it was written before, then it would have been incorrect
-- because there were only two addarg() calls total.  The index in the loop() must match the
-- number of addarg() calls.
    test_parser:loop(3)
    test_parser:nofiles()

    local testB = clink.argmatcher()
    testB:addflags("-flagA", "-flagB", "-optionC")
    testB:addarg({ "hey", "this",  "isa", "test" })
    testB:nofiles()

    --------------------------------------------------------------------------------

    local root_parser = clink.argmatcher("t")
    root_parser:addarg(
        {
            "test_parser" .. test_parser,
            "test-B" .. testB
        }
    )
```

**Comment by chrisant996** [comment]
Hey @ben-may , FWIW in v1.3.11 while I was adding back support for merging argmatchers* I discovered there was a regression in `set_arguments()` and `add_arguments()` when passing multiple argument position definitions.  E.g. when adding 3 separate argument definitions (such as `{ {"first"}, {"second"}, {"third"} }`), all of the positions were combined into a single argument position.

That's fixed in v1.3.11, and also argmatcher merging works even better now than it did in v0.4.9 (neither is perfect, and neither tries to be perfect, but the ability is both restored and improved).

`*` Merging argmatchers was lost in Martin's v1.0.0 rewrite, and in v1.1.4 I had added back some very limited support for merging.

**Comment by ben-may** [comment]
This is awesome and would help with backwards compatibility issues I ran into when upgrading!  Thanks!
